File: src/main/java/pe/nanamochi/banchus/entities/commons/Mods.java
package pe.nanamochi.banchus.entities.commons;

import java.util.*;
import java.util.stream.Collectors;
import lombok.AllArgsConstructor;
import lombok.Getter;

@Getter
@AllArgsConstructor
public enum Mods {
  NO_MOD(0, "No Mod", "NM"),
  NO_FAIL(1, "No Fail", "NF"),
  EASY(1 << 1, "Easy", "EZ"),
  NO_VIDEO(1 << 2, "No Video", "NV"), // Replaced by "Toushscreen" in later versions
  HIDDEN(1 << 3, "Hidden", "HD"),
  HARD_ROCK(1 << 4, "Hard Rock", "HR"),
  SUDDEN_DEATH(1 << 5, "Sudden Death", "SD"),
  DOUBLE_TIME(1 << 6, "Double Time", "DT"),
  RELAX(1 << 7, "Relax", "RX"),
  HALF_TIME(1 << 8, "Half Time", "HT"),
  NIGHTCORE(1 << 9, "Nightcore", "NC"), // Used as "Taiko" mod in older versions
  FLASHLIGHT(1 << 10, "Flashlight", "FL"),
  AUTOPLAY(1 << 11, "Autoplay", "AU"),
  SPUN_OUT(1 << 12, "Spun Out", "SO"),
  AUTOPILOT(1 << 13, "Autopilot", "AP"),
  PERFECT(1 << 14, "Perfect", "PF"),
  KEY4(1 << 15, "4K", "4K"),
  KEY5(1 << 16, "5K", "5K"),
  KEY6(1 << 17, "6K", "6K"),
  KEY7(1 << 18, "7K", "7K"),
  KEY8(1 << 19, "8K", "8K"),
  FADE_IN(1 << 20, "Fade-In", "FI"),
  RANDOM(1 << 21, "Random", "RD"),
  CINEMA(1 << 22, "Cinema", "CN"),
  TARGET(1 << 23, "Target Practice", "TC"),
  KEY9(1 << 24, "9K", "9K"),
  KEY_COOP(1 << 25, "Co-Op", "CO"),
  KEY1(1 << 26, "1K", "1K"),
  KEY3(1 << 27, "3K", "3K"),
  KEY2(1 << 28, "2K", "2K"),
  SCORE_V2(1 << 29, "Score V2", "V2"),
  MIRROR(1 << 30, "Mirror", "MR"),
  ;

  public static final int SPEED_CHANGING =
      DOUBLE_TIME.getValue() | HALF_TIME.getValue() | NIGHTCORE.getValue();
  private static final int KEY_MODS_MASK =
      KEY1.value
          | KEY2.value
          | KEY3.value
          | KEY4.value
          | KEY5.value
          | KEY6.value
          | KEY7.value
          | KEY8.value
          | KEY9.value
          | KEY_COOP.value;
  public static final int OSU_SPECIFIC_MODS = SPUN_OUT.value | AUTOPILOT.value;
  private static final int MANIA_SPECIFIC_MODS =
      RANDOM.value | MIRROR.value | FADE_IN.value | KEY_MODS_MASK;

  private final int value;
  private final String displayName;
  private final String initial;

  public static List<Mods> fromBitmask(int bit) {
    if (bit == 0) return new ArrayList<>(List.of(NO_MOD));

    return Arrays.stream(values())
        .filter(mod -> mod != NO_MOD && (bit & mod.value) == mod.value)
        .collect(Collectors.toList());
  }

  public static List<Mods> fromInitials(String[] initials) {
    return Arrays.stream(values())
        .filter(
            mod -> {
              for (String initial : initials) {
                if (mod.getInitial().equalsIgnoreCase(initial)) {
                  return true;
                }
              }
              return false;
            })
        .collect(Collectors.toList());
  }

  public static int toBitmask(List<Mods> mods) {
    if (mods == null) return 0;
    return mods.stream().mapToInt(Mods::getValue).reduce(0, (a, b) -> a | b);
  }

  public static int filterInvalidModCombinations(int bitmask, Mode mode) {
    int result = bitmask;

    if ((result & (DOUBLE_TIME.value | NIGHTCORE.value)) == (DOUBLE_TIME.value | NIGHTCORE.value)) {
      result &= ~DOUBLE_TIME.value;
    }

    if (((result & (DOUBLE_TIME.value | NIGHTCORE.value)) != 0)
        && (result & HALF_TIME.value) != 0) {
      result &= ~HALF_TIME.value;
    }

    if ((result & EASY.value) != 0 && (result & HARD_ROCK.value) != 0) {
      result &= ~HARD_ROCK.value;
    }

    if ((result & (NO_FAIL.value | RELAX.value | AUTOPILOT.value)) != 0) {
      result &= ~SUDDEN_DEATH.value;
      result &= ~PERFECT.value;
    }

    if ((result & (RELAX.value | AUTOPILOT.value)) != 0) {
      result &= ~NO_FAIL.value;
    }

    if ((result & PERFECT.value) != 0 && (result & SUDDEN_DEATH.value) != 0) {
      result &= ~SUDDEN_DEATH.value;
    }

    if (mode.getValue() != Mode.OSU.getValue()) {
      result &= ~OSU_SPECIFIC_MODS;
    }

    if (mode.getValue() != Mode.MANIA.getValue()) {
      result &= ~MANIA_SPECIFIC_MODS;
    }

    if (mode.getValue() == Mode.OSU.getValue()) {
      if ((result & AUTOPILOT.value) != 0) {
        if ((result & (SPUN_OUT.value | RELAX.value)) != 0) {
          result &= ~AUTOPILOT.value;
        }
      }
    }

    if (mode.getValue() == Mode.MANIA.getValue()) {
      result &= ~RELAX.value;

      if ((result & HIDDEN.value) != 0 && (result & FADE_IN.value) != 0) {
        result &= ~FADE_IN.value;
      }
    }

    int keyModsUsed = result & KEY_MODS_MASK;
    if (Integer.bitCount(keyModsUsed) > 1) {
      for (Mods mod : Mods.values()) {
        if ((mod.value & KEY_MODS_MASK) != 0 && (keyModsUsed & mod.value) != 0) {
          result &= ~(keyModsUsed & ~mod.value);
          break;
        }
      }
    }

    return result;
  }

  public static List<Mods> filterInvalidModCombinations(List<Mods> mods, Mode mode) {
    int bitmask = toBitmask(mods);
    int cleanBitmask = filterInvalidModCombinations(bitmask, mode);
    return fromBitmask(cleanBitmask);
  }

  public static Mods getManiaKeyCount(List<Mods> mods) {
    if (mods.contains(KEY9)) {
      return KEY9;
    } else if (mods.contains(KEY8)) {
      return KEY8;
    } else if (mods.contains(KEY7)) {
      return KEY7;
    } else if (mods.contains(KEY6)) {
      return KEY6;
    } else if (mods.contains(KEY5)) {
      return KEY5;
    } else if (mods.contains(KEY4)) {
      return KEY4;
    } else if (mods.contains(KEY3)) {
      return KEY3;
    } else if (mods.contains(KEY2)) {
      return KEY2;
    }
    return KEY1;
  }
}

File: banchus-app/src/main/kotlin/pe/nanamochi/banchus/core/enums/Mods.kt
package pe.nanamochi.banchus.core.enums

enum class Mods(val value: UInt, val displayName: String, val initial: String) {
    NO_MOD(0u, "No Mod", "NM"),
    NO_FAIL(1u shl 0, "No Fail", "NF"),
    EASY(1u shl 1, "Easy", "EZ"),
    NO_VIDEO(1u shl 2, "No Video", "NV"),
    HIDDEN(1u shl 3, "Hidden", "HD"),
    HARD_ROCK(1u shl 4, "Hard Rock", "HR"),
    SUDDEN_DEATH(1u shl 5, "Sudden Death", "SD"),
    DOUBLE_TIME(1u shl 6, "Double Time", "DT"),
    RELAX(1u shl 7, "Relax", "RX"),
    HALF_TIME(1u shl 8, "Half Time", "HT"),
    NIGHTCORE(1u shl 9, "Nightcore", "NC"),
    FLASHLIGHT(1u shl 10, "Flashlight", "FL"),
    AUTOPLAY(1u shl 11, "Autoplay", "AU"),
    SPUN_OUT(1u shl 12, "Spun Out", "SO"),
    AUTOPILOT(1u shl 13, "Autopilot", "AP"),
    PERFECT(1u shl 14, "Perfect", "PF"),
    KEY4(1u shl 15, "4K", "4K"),
    KEY5(1u shl 16, "5K", "5K"),
    KEY6(1u shl 17, "6K", "6K"),
    KEY7(1u shl 18, "7K", "7K"),
    KEY8(1u shl 19, "8K", "8K"),
    FADE_IN(1u shl 20, "Fade-In", "FI"),
    RANDOM(1u shl 21, "Random", "RD"),
    CINEMA(1u shl 22, "Cinema", "CN"),
    TARGET(1u shl 23, "Target Practice", "TC"),
    KEY9(1u shl 24, "9K", "9K"),
    KEY_COOP(1u shl 25, "Co-Op", "CO"),
    KEY1(1u shl 26, "1K", "1K"),
    KEY3(1u shl 27, "3K", "3K"),
    KEY2(1u shl 28, "2K", "2K"),
    SCORE_V2(1u shl 29, "Score V2", "V2"),
    MIRROR(1u shl 30, "Mirror", "MR");

    companion object {
        val SPEED_CHANGING: UInt = DOUBLE_TIME.value or HALF_TIME.value or NIGHTCORE.value

        val KEY_MODS_MASK: UInt =
            KEY1.value or
                KEY2.value or
                KEY3.value or
                KEY4.value or
                KEY5.value or
                KEY6.value or
                KEY7.value or
                KEY8.value or
                KEY9.value or
                KEY_COOP.value

        val OSU_SPECIFIC_MODS: UInt = SPUN_OUT.value or AUTOPILOT.value

        val MANIA_SPECIFIC_MODS: UInt =
            RANDOM.value or MIRROR.value or FADE_IN.value or KEY_MODS_MASK

        fun fromBitmask(bitmask: UInt): List<Mods> {
            if (bitmask == 0u) return listOf(NO_MOD)
            return entries.filter { it != NO_MOD && (bitmask and it.value) == it.value }
        }

        fun fromInitials(initials: Array<String>): List<Mods> {
            return entries.filter { mod ->
                initials.any { it.equals(mod.initial, ignoreCase = true) }
            }
        }

        fun toBitmask(mods: List<Mods>?): UInt =
            mods?.fold(0u) { acc, mod -> acc or mod.value } ?: 0u

        fun hasConflict(bitmask: UInt): Boolean {
            if ((bitmask and DOUBLE_TIME.value) != 0u && (bitmask and HALF_TIME.value) != 0u) {
                return true
            }

            if ((bitmask and NIGHTCORE.value) != 0u && (bitmask and DOUBLE_TIME.value) == 0u) {
                return true
            }

            if ((bitmask and EASY.value) != 0u && (bitmask and HARD_ROCK.value) != 0u) {
                return true
            }

            if ((bitmask and RELAX.value) != 0u && (bitmask and AUTOPILOT.value) != 0u) {
                return true
            }

            if ((bitmask and HIDDEN.value) != 0u && (bitmask and FADE_IN.value) != 0u) {
                return true
            }

            if ((bitmask and KEY_MODS_MASK).countOneBits() > 1) {
                return true
            }

            return false
        }

        fun filterInvalidModCombinations(bitmask: UInt, mode: Mode): UInt {
            var result = bitmask

            if (
                (result and (DOUBLE_TIME.value or NIGHTCORE.value)) ==
                    (DOUBLE_TIME.value or NIGHTCORE.value)
            ) {
                result = result and DOUBLE_TIME.value.inv()
            }

            if (
                ((result and (DOUBLE_TIME.value or NIGHTCORE.value)) != 0u) &&
                    (result and HALF_TIME.value) != 0u
            ) {
                result = result and HALF_TIME.value.inv()
            }

            if ((result and EASY.value) != 0u && (result and HARD_ROCK.value) != 0u) {
                result = result and HARD_ROCK.value.inv()
            }

            if ((result and (NO_FAIL.value or RELAX.value or AUTOPILOT.value)) != 0u) {
                result = result and SUDDEN_DEATH.value.inv()
                result = result and PERFECT.value.inv()
            }

            if ((result and (RELAX.value or AUTOPILOT.value)) != 0u) {
                result = result and NO_FAIL.value.inv()
            }

            if ((result and PERFECT.value) != 0u && (result and SUDDEN_DEATH.value) != 0u) {
                result = result and SUDDEN_DEATH.value.inv()
            }

            if (mode != Mode.OSU) {
                result = result and OSU_SPECIFIC_MODS.inv()
            }

            if (mode != Mode.MANIA) {
                result = result and MANIA_SPECIFIC_MODS.inv()
            } else {
                result = result and RELAX.value.inv()
                if ((result and HIDDEN.value) != 0u && (result and FADE_IN.value) != 0u) {
                    result = result and FADE_IN.value.inv()
                }
            }

            if (mode == Mode.OSU) {
                if (
                    (result and AUTOPILOT.value) != 0u &&
                        (result and (SPUN_OUT.value or RELAX.value)) != 0u
                ) {
                    result = result and AUTOPILOT.value.inv()
                }
            }

            val keyModsUsed = result and KEY_MODS_MASK
            if (keyModsUsed.countOneBits() > 1) {
                entries
                    .find { (it.value and KEY_MODS_MASK) != 0u && (keyModsUsed and it.value) != 0u }
                    ?.let { first -> result = result and (keyModsUsed and first.value.inv()).inv() }
            }

            return result
        }

        fun filterInvalidModCombinations(mods: List<Mods>, mode: Mode): List<Mods> {
            val bitmask = toBitmask(mods)
            val cleanBitmask = filterInvalidModCombinations(bitmask, mode)
            return fromBitmask(cleanBitmask)
        }

        fun getManiaKeyCount(mods: List<Mods>): Mods {
            val keyPriority = listOf(KEY9, KEY8, KEY7, KEY6, KEY5, KEY4, KEY3, KEY2)
            return keyPriority.find { it in mods } ?: KEY1
        }
    }
}
